rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java

```diff
@@ -0,0 +1,897 @@
+        }).toArray(URI[]::new);
+
+        // check that the multiLevelNestedActivity method returned the mandatory LRA followed by any nested LRAs
+        assertEquals("multiLevelNestedActivity: step 1 (the test call went to " + resourcePath.getUri() + ")",
+                nestedCnt + 1, uris.length);
+        // first element should be the mandatory LRA
+        assertEquals("multiLevelNestedActivity: step 2 (the test call went to " + resourcePath.getUri() + ")",
+                lra, uris[0]);
+
+        // and the mandatory lra seen by the multiLevelNestedActivity method
+        assertFalse("multiLevelNestedActivity: top level LRA should be active (path called " + resourcePath.getUri() + ")",
+                isFinished(uris[0]));
+
+//        lraTestService.waitForCallbacks(lra);
+
+        // check that all nested activities were told to complete
+        assertEquals("multiLevelNestedActivity: step 3 (called test path " +
+                resourcePath.getUri() + ")", nestedCnt, completeCount.get());
+        assertEquals("multiLevelNestedActivity: step 4 (called test path " +
+                resourcePath.getUri() + ")", 0, compensateCount.get());
+
+        // close the LRA
+        if (how == CompletionType.compensate) {
+            lraClient.cancelLRA(lra);
+
+            // validate that the top level and nested LRAs are gone
+            assertAllFinished(uris);
+            /*
+             * the test starts LRA1 calls a @Mandatory method multiLevelNestedActivity which enlists in LRA1
+             * multiLevelNestedActivity then calls an @Nested method which starts L2 and enlists another participant
+             *   when the method returns the nested participant is completed (ie completed count is incremented)
+             * Canceling L1 should then compensate the L1 enlistment (ie compensate count is incremented)
+             * which will then tell L2 to compensate (ie the compensate count is incremented again)
+             */
+            // each nested participant should have completed (the +nestedCnt)
+            assertEquals("multiLevelNestedActivity: step 7 (called test path " +
+                    resourcePath.getUri() + ")", nestedCnt, completeCount.get());
+            // each nested participant should have compensated. The top level enlistment should have compensated (the +1)
+            assertEquals("multiLevelNestedActivity: step 8 (called test path " +
+                    resourcePath.getUri() + ")", nestedCnt + 1, compensateCount.get());
+        } else if (how == CompletionType.complete) {
+            lraClient.closeLRA(lra);
+
+            // validate that the top level and nested LRAs are gone
+            assertAllFinished(uris);
+            // make sure that all nested activities were not told to complete or cancel a second time
+            assertEquals("multiLevelNestedActivity: step 5 (called test path " +
+                    resourcePath.getUri() + ")", 2 * nestedCnt, completeCount.get());
```

Comment:
> @mmusgrov as I'm looking more into the changes I have to admit I don't understand clearly on the relationship of the newly added `LRAParentAbstractRecord`, `LRAChildAbstractRecord` and how it changes the current behaviour.
> Would you be so kind and elaborate a bit on this from design and functionality perspective?
> Thank you.

When the parent and child are in the same VM we make better use of ArjunaCore and avoid some of the JAX-RS calls by making in-VM calls instead.

And it has architectural benefits as well:

- it sets up the code for further enhancements in how we handle remote coordinators where we can more closely follow other models such as JTS;
- it lays the ground for using ArjunaCore in ServerLRAFilter to reduce the possibility of hitting FailedToXXX states (see JBTM-3245). That will require a persistent transaction store in each microservice (currently we only require such a store in the coordinator). Such a store will also allow microservices to store state which will enable them to more easily fulfil participant requirements - note that LRA is an optimistic protocol (if you recall it was originally called JFDI) but such optimism introduces the potential for reaching failed states.

I don't want us to re-architect it now to meet all of those goals but I'd like someone to do it in the future.

Reply:
It is quite a complicated test so I will need to take a more in depth look at the mechanics of the test. I will update this conversation when we have reached agreement on the other outstanding conversations in the PR.

Reply:
So the assertion needs to validate that each nested (participant) was asked to complete and that the top level (participant) was asked to complete also. That makes a total of at least `nestedCnt + 1` completion notifications. So the test only worked because nestedCnt has the value 1 (and for any other value it would have failed). I updated the assertion to verify that at least `nestedCnt + 1` completion notification were received (and I also added an assertion for the forget count for good measure). Thanks for the catch on such a tricky test :-)